### PR TITLE
Enable source maps in webpack chunking + bundling process

### PR DESCRIPTION
**Pull request by ptomasroos**
Its impossible to get good source maps at this moment in next.js
The reason for this is that the "manual" concat of sources happens outside the sourcemap generation which ends up resulting in, guess what! No source maps for app.js (which is quite crucial because it often exists of majority of the application)

```
➜  web git:(fix-850-source-maps) ✗ ls -l .next 
total 12552
-rw-r--r--  1 tomas  staff       36 Feb  9 09:43 BUILD_ID
-rw-r--r--  1 tomas  staff  1195881 Feb  9 09:43 app.js
-rw-r--r--  1 tomas  staff       54 Feb  9 09:43 build-stats.json
drwxr-xr-x  3 tomas  staff       96 Feb  9 09:43 bundles
-rw-r--r--  1 tomas  staff  1063099 Feb  9 09:43 commons.js
-rw-r--r--  1 tomas  staff  3728876 Feb  9 09:43 commons.js.map
drwxr-xr-x  4 tomas  staff      128 Feb  9 09:43 dist
-rw-r--r--  1 tomas  staff    29961 Feb  9 09:43 main.js
-rw-r--r--  1 tomas  staff   117438 Feb  9 09:43 main.js.map
-rw-r--r--  1 tomas  staff     3135 Feb  9 09:43 manifest.js
-rw-r--r--  1 tomas  staff    10791 Feb  9 09:43 manifest.js.map
-rw-r--r--  1 tomas  staff    99826 Feb  9 09:43 react.js
-rw-r--r--  1 tomas  staff   124422 Feb  9 09:43 react.js.map
-rw-r--r--  1 tomas  staff    25859 Feb  9 09:43 service-worker.js
```

This is sad of course! 

This PR aims to fix this by removing the home written source concat and using CommonChunksPlugin to combine main, manifest (added to last defined common chunk, this is the bootstrap code of webpack), react and the commons stuff.

In order to use source-maps one needs to alter the webpack config through next.config and enable it with the following 
```
      config.devtool = 'source-map';

      // Perform customizations to config existing plugins
      for (const options of config.plugins) {
        if (options['constructor']['name'] === 'UglifyJsPlugin') {
          options.options.sourceMap = true;
          break;
        }
      }
```

First set webpack's devtool to 'source-map' in order to get external maps.

And then we're using Uglify and there we need to find the plugin and override the default sourceMap: false from "user-land" config.

```
total 10424
-rw-r--r--  1 tomas  staff       36 Feb 13 21:05 BUILD_ID
-rw-r--r--  1 tomas  staff  1203400 Feb 13 21:05 app.js
-rw-r--r--  1 tomas  staff  4094606 Feb 13 21:05 app.js.map
-rw-r--r--  1 tomas  staff       54 Feb 13 21:05 build-stats.json
drwxr-xr-x  3 tomas  staff       96 Feb 13 21:05 bundles
drwxr-xr-x  4 tomas  staff      128 Feb 13 21:05 dist
-rw-r--r--  1 tomas  staff    25859 Feb 13 21:05 service-worker.js
```

### CODE REVIEW

File: server/build/webpack.js
Comment (timneutkens) [inline server/build/webpack.js]: This negates the react split bundle optimization with uglify. It might not be needed anymore because of the uglify optimizations we did though. cc @arunoda

File: server/build/webpack.js
Comment (arunoda) [inline server/build/webpack.js]: Remove this completely.

Comment (arunoda) [inline server/build/webpack.js]: Remove this completely. We are removing the react bundle too.
So, there's no need this in `dev`.

File: server/build/webpack.js
Comment (arunoda) [reply server/build/webpack.js]: @ptomasroos let's remove this as well.

File: server/build/webpack.js
Comment (arunoda) [inline server/build/webpack.js]: Let's keep this.
This helps in bundle rebuilding process.

**Review by arunoda — CHANGES_REQUESTED** [review]
Yep. I like this approach.
Have a look at my suggestions below.

File: server/build/webpack.js
Comment (arunoda) [inline server/build/webpack.js]: We can keep this as `main.js`.
And in the `server/document.js` we can change to use `main.js` instead of `app.js` for both production and dev.

For dev you may need to use `manifest.js` before `main.js`.

**Review by arunoda — CHANGES_REQUESTED** [review]
have a look at these suggestions too.
After that, we may've to few other changes.

I'll send a patch for that.

**Comment by ptomasroos** [comment]
All right @arunoda I've fixed what you mentioned and also cleaned up the fix around late bundling regarding the build-stats

File: server/build/webpack.js
Comment (ptomasroos) [reply server/build/webpack.js]: Done

File: server/build/webpack.js
Comment (ptomasroos) [reply server/build/webpack.js]: Done

File: server/build/webpack.js
Comment (ptomasroos) [reply server/build/webpack.js]: Done

**Comment by ptomasroos** [comment]
Ping @arunoda

**Comment by nilportugues** [comment]
Hey, can we get this merged ASAP? 

Ping @arunoda

**Comment by juanramon** [comment]
Ping @arunoda

**Comment by arunoda** [comment]
@ptomasroos we were waiting because of our Webpack v4 migration. Seems like we need to wait a bit more for it.

Let me fix conflicts on this.

**Comment by timneutkens** [comment]
@nilportugues we always thoroughly verify all PRs changing default behaviour. So sometimes it takes a while.

@juanramon it doesn't make sense to send `ping` twice in a hour. It won't make a maintainer merge the pull request any faster, and might even be viewed as aggressive by some people (not us), just a small advice 👍

🙏

**Comment by ptomasroos** [comment]
Thanks @arunoda feel free to fix the conflicts

**Review by arunoda — APPROVED** [review]

**Comment by arunoda** [comment]
@ptomasroos I've fixed the conflict and make sure tests are passing.
Have look at the changes.

@timneutkens could you do a final review and take this in.

**Comment by ptomasroos** [comment]
Thanks @arunoda all looks good

**Review by timneutkens — APPROVED** [review]
Looks great, cleans up a lot of code too 😌

**Comment by devcorpio** [comment]
Hi @timneutkens,

Only for know, how much time usually do you wait before merge canary branch into master? With this change will begin to appear the map of the file app.js?

Regards,

**Comment by timneutkens** [comment]
We're currently finishing up for 5.1. No ETA yet, but this has been release as `next@canary` already 👍

**Comment by oliviertassinari** [comment]
I'm happy to see the following diff :):
```diff
-app.js
+main.js
```

**Comment by timneutkens** [comment]
Amazing 😲

**Comment by devcorpio** [comment]
Then, from now, nextjs will be generate main.js(and main.js.map) instead of app.js?

That change will be helpful for my workmates and me, will help us using sentry reporting tool :)

**Comment by timneutkens** [comment]
Yep @devcorpio

**Comment by bryandowning** [comment]
Any chance this would have messed with `@zeit/next-sass`? My sass doesn't compile correctly after upgrading from `next@5.0.1-canary.10` to `next@5.0.1-canary.11`. I'm having a hard time reproducing a simple example. It's weird because some of the styles get output correctly, but other component styles are omitted.

**Comment by timneutkens** [comment]
@bryandowning most likely you're locked to an older version of next-css. Try removing the lockfiles, removing node_modules and reinstalling and upgrading next-sass. I'll do a new release soon that locks all versions of next-plugins, then this won't be happening again, because it'll just lock to the version specified.

**Pull request merged**
